apps/src/dropletUtils.js

```diff
@@ -404,7 +417,15 @@ exports.generateAceApiCompleter = function (functionFilter, dropletConfig) {
         callback(null, []);
         return;
       }
-      callback(null, apis);
+      var line = session.getLine(pos.row);
+      var identifier = acUtil.retrievePrecedingIdentifier(line, pos.column);
+      if (pos.column > (identifier.length + 1) &&
+          line[pos.column - identifier.length - 1] === '.') {
```

Comment:
Not sure there is an error. I intentionally had it go in the else case if the line begins with a dot. I wanted it to only go into the new methods completer if there is both a dot and something before the dot before the current identifier. If the line is too short to have something before the dot, then we do what we also did in the past. I'll make sure I add a comment. And can extract a helper to improve readability.

The other comment needed is that we're not doing a sophisticated parse of the whole document - ace autocomplete is much more simplistic. So multiline expressions won't get autocomplete.

We may eventually want to examine the identifier preceding the dot more carefully. I have a note to try to avoid autocompleting `1.` since you're probably typing a decimal. But technically `(1).` could be autocompleted with `toString()`, so this is a little tricky. 

But thanks for examining in such detail regardless!



> On Mar 1, 2016, at 4:32 PM, Brad Buchanan <notifications@github.com> wrote:
> 
> In apps/src/dropletUtils.js:
> 
> > @@ -404,7 +417,15 @@ exports.generateAceApiCompleter = function (functionFilter, dropletConfig) {
> >          callback(null, []);
> >          return;
> >        }
> > -      callback(null, apis);
> > +      var line = session.getLine(pos.row);
> > +      var identifier = acUtil.retrievePrecedingIdentifier(line, pos.column);
> > +      if (pos.column > (identifier.length + 1) &&
> > +          line[pos.column - identifier.length - 1] === '.') {
> Grabbed @Hamms to help me parse this, and he pointed out that the second condition is correct and checks whether the cursor is inside a symbol that immediately follows a dot.
> 
> We eventually tracked down a failing case: a line that begins with . will incorrectly take the else case even when the cursor is in such a position. That's due to an off-by-one error in the first condition; which, incidentally, is unneeded since indexing outside of a string will just return undefined.
> 
> But that off-by-one error took way too much work to find. Let's just extract and test a isCursorInSymbolAfterDot helper, please?
> 
> —
> Reply to this email directly or view it on GitHub.
>